onnxruntime/core/framework/session_state.cc

```diff
@@ -1196,6 +1203,17 @@ static Status VerifyEachNodeIsAssignedToAnEp(const Graph& graph, const logging::
   }
 #endif  // !defined(ORT_MINIMAL_BUILD)
 
+  // Silent fallback from GPU/NPU to CPU nodes can cause performance issues due to memory copies and frequent stalls.
+  // If the user explicitly included the CPU provider anyway, then remain silent, but if it was implicitly added,
+  // and unexpected fallback happened to a non-preferred provider, warn the user.
+  size_t explicit_provider_count = providers.NumProviders() - providers.GetCpuProviderWasImplicitlyAdded() ? 1 : 0;
+  if (node_placement_set.size() > explicit_provider_count) {
+    LOGS(logger, WARNING) << "Some nodes were not assigned to the preferred execution providers which could decrease performance.";
```

Comment:
Like we discussed some time ago not all placements on the non-preferred provider has a negative impact on performance; for shape (for e.g.) related operations it's beneficial to run them on CPU. Can we soften this message a bit? ```Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. For e.g. ORT explicitly assigns shape related ops to CPU to improve perf```.

Reply:
Sure. Pending.